### PR TITLE
Issue #5962: Testing README clarifications

### PR DESCRIPTION
**Pull request by ncoghlan**
- consolidate dependency installation instructions to one place
- reference HACKING.md from test/README
- make `testsuite-prepare` more prominent
- note that `rmmod` error is harmless
- reference Fedora wiki for nested virtualization problems

### CODE REVIEW

File: test/README [hunk 1]
```diff
@@ -3,33 +3,49 @@
...
 ## Introduction
 
-To run the integration tests run the following. Don't run the integration tests
-as root:
+Before running the tests, ensure Cockpit has been built where the test suite
+expects to find it (do NOT run the build step as root):
+
+   $ ./test/verify/testsuite-prepare
+
+Only code currently committed in the current branch will be installed and tested.
```
Comment (mvollmer) [inline test/README]: This is no longer true, but that's a separate issue.

File: test/README [hunk 2]
```diff
@@ -24,17 +24,14 @@ for more details and recommendations on ensuring it is enabled correctly.
 
 ## Introduction
 
-Before running the tests, ensure Cockpit has been built where the test suite
-expects to find it (do NOT run the build step as root):
+To build Cockpit and run the integration tests run the following (do NOT run
+the integration tests as root):
 
-   $ ./test/verify/testsuite-prepare
+    $ ./test/verify/run-tests --install
```
Comment (martinpitt) [inline test/README]: I'm not sure I like that. I would like to keep documenting it as two separate steps. It's unnerving and absolutely unnecessary to rebuild the test VM (i. e. the `testsuite-prepare` part) when working on the tests, i. e. the installed cockpit doesn't change. It's of course fine to point out that running `run-tests` or any `check-*` with `--install` is a shortcut for this step.

File: HACKING.md [hunk 3]
```diff
@@ -26,6 +28,46 @@ And lastly get Webpack and the development dependencies:
...
+        openssl \
+        python \
+        python-libguestfs \
+        python-lxml \
+        qemu \
+        qemu-kvm \
+        rpm-build \
+        selinux-policy-devel \
+        rsync \
+        xz
```
Comment (martinpitt) [inline HACKING.md]: This is hard to copy&paste and unnecessarily long. Can you please reformat this to just use two or three lines?

File: HACKING.md [hunk 4]
```diff
@@ -26,6 +28,46 @@ And lastly get Webpack and the development dependencies:
...
+When relying on CI to run the test suite, this is all that is
+necessary to work on the JavaScript components of Cockpit.
+
+To actually build the Cockpit binaries themselves from source
+(including to run the integration tests locally), you will need
+additional header files and other components. Check
+`tools/cockpit.spec` for the concrete Fedora build dependencies.
+The following should work in a fresh Git clone:
+
+    $ sudo yum-builddep tools/cockpit.spec
```
Comment (martinpitt) [inline HACKING.md]: This is missing node and npm; that part got removed below, but not re-added here.

File: HACKING.md [hunk 5]
```diff
@@ -26,6 +28,46 @@ And lastly get Webpack and the development dependencies:
...
+        curl \
+        expect \
+        firewalld \
+        krb5-workstation \
+        krb5-server \
+        libvirt \
+        libvirt-client \
+        libvirt-daemon \
+        libvirt-python \
+        mock \
```
Comment (martinpitt) [inline HACKING.md]: I don't think we should add mock here (I know, it was in the original list already). You can (and should) use mock for building stuff, but it's not cockpit specific, and it's unrelated to running the tests.

File: test/README [hunk 6]
```diff
@@ -3,33 +3,49 @@
...
+will handle that.
 
 You may need to reboot after installing the above packages. For initialization
 the following command needs to be run as root to setup the libvirt bridged network:
 
     $ sudo ./test/vm-prep
 
+Note that this step also attempts to ensure nested virtualization is enabled,
+and may emit a spurious `rmmod: ERROR: Module kvm_amd is not currently loaded`
+warning when running on an Intel CPU (or vice-versa).
```
Comment (martinpitt) [inline test/README]: I wouldn't document this -- I'd rather just fix the script to not show that error in the first place.

File: HACKING.md [hunk 7]
```diff
@@ -26,6 +28,40 @@ And lastly get Webpack and the development dependencies:
...
+        libvirt-daemon \
+        libvirt-python \
+        python \
+        python-libguestfs \
+        python-lxml \
+        qemu \
+        qemu-kvm \
+        rpm-build \
+        rsync \
+        xz
```
Comment (martinpitt) [inline HACKING.md]: This is hard to copy&paste and unnecessarily long. Can you please reformat this to just use two or three lines?

File: HACKING.md (hunk 4 above)
Comment (mvollmer) [inline HACKING.md]: nodejs and npm are explicitly installed before this paragraph, no?

File: test/README (hunk 1 above)
Comment (ncoghlan) [inline test/README]: Do you want me to drop that paragraph?

File: HACKING.md (hunk 7 above)
Comment (ncoghlan) [inline HACKING.md]: Done!

File: HACKING.md (hunk 4 above)
Comment (ncoghlan) [inline HACKING.md]: Right, nodejs, npm, and webpack are installed above, so this is just the extra dependencies to build the C components.

File: test/README (hunk 6 above)
Comment (ncoghlan) [inline test/README]: While I agree in principle, perhaps add it for now, and then take it out again when the script is fixed?

I didn't try to change the behaviour myself, as I'm not clear on the best way to go - the script has an apparently flawed assumption that "AMD/Intel KVM config file present -> system is running that kind of chip", but I don't know whether it should be addressed by finding a better query condition, or else by just ignoring the error when it happens.

File: test/README (hunk 1 above)
Comment (mvollmer) [inline test/README]: Yes, why not.  Thanks!

File: HACKING.md (hunk 4 above)
Comment (martinpitt) [inline HACKING.md]: Oh right, didn't look into the expander above, sorry for the noise.

File: test/README (hunk 6 above)
Comment (martinpitt) [inline test/README]: Perhaps something like this:

    [ -d /sys/modules/kvm_intel ] && rmmod kvm-intel && modprobe kvm-intel || true

and same for amd? that should avoid the noise and still show relevant errors.

This is aside from the fact that these modprobe.d snippets don't actually seem to work -- I need to boot with `kvm-intel.nested=1` on the kernel command line to really enable nested kvm.

File: test/README [hunk 8]
```diff
@@ -3,33 +3,46 @@
...
-as root:
+To build Cockpit and run the integration tests run the following (do NOT run
+the integration tests as root):
 
-    $ ./test/verify/run-tests
+    $ ./test/verify/run-tests --install
+
+The `--install` option indicates to (re)build and (re)install Cockpit before
+running the tests. Only code currently committed in the current branch will
+be installed and tested.
```
Comment (martinpitt) [inline test/README]: I'm not sure I like that. I would like to keep documenting it as two separate steps. It's unnerving and absolutely unnecessary to rebuild the test VM (i. e. the `testsuite-prepare` part) when working on the tests, i. e. the installed cockpit doesn't change. It's of course fine to point out that running `run-tests` or any `check-*` with `--install` is a shortcut for this step.

(Already commened on this before, but it got hidden under "obsolete comment").

**Comment by stefwalter** [comment]
@martinpitt This is the sort of pull request where it would be helpful if you could post a commit or FIXUP a commit. Thanks @ncoghlan for the clarifications. For this stuff, it's super helpful to have the insight of someone who isn't steeped in their Cockpit development workflow.

File: test/README (hunk 6 above)
Comment (ncoghlan) [inline test/README]: I created a separate PR for this: https://github.com/cockpit-project/cockpit/pull/5981

So I'll drop this paragraph from the README update.

**Comment by ncoghlan** [comment]
OK, I think the last couple of commits and the separate PR for #5981 address @martinpitt's remaining comments.

**Pull request closed without merging**